### PR TITLE
[ENG-3896] POST functionality on ResourceList endpoint

### PR DESCRIPTION
**Pull request by jwalz**
## Purpose
Implement POST functionality for the top-level `/v2/resources/` endpoint.

POST should accept a `registration` relationship that is used to determine permissions and identify the relevant Outcome to add a blank Resource. All Resource metadata will be provided by follow-on PATCH requests to `/v2/resources/<resource_id>`

## Changes
* Add ResourceList View and route to it via `/v2/resources`
* Add new Permissions class for ResourceList
* Add new `create` functionality to ResourceSerializer
* Tests

## QA Notes
For reasons utterly alien to me,  likely having to do with `primary_resource_guid` being populated via annotation, the ResourceSerializer is not loading the target Registration into `validated_data`, so we are bucking convention and getting it straight from the request.

## Ticket

https://openscience.atlassian.net/browse/ENG-3896

### CODE REVIEW

File: api/resources/serializers.py [hunk 1]
```diff
@@ -60,3 +62,15 @@ def get_absolute_url(self, obj):
...
+
+    def create(self, validated_data):
+        # Already loaded by view, so no need to error check
+        guid = self.context['request'].data['registration']
+        primary_registration = Registration.load(guid)
+
+        try:
+            root_outcome = Outcome.objects.for_registration(primary_registration, create=True)
+        except NoPIDError:
+            raise Conflict('Cannot add Resrouces to a Registration without a DOI')
```
Comment (brianjgeiger) [inline api/resources/serializers.py:74]:
```suggestion
            raise Conflict('Cannot add Resources to a Registration that does not have a DOI')
```

Two things here: Resources was misspelled, and I'm guessing the registration is the thing that needs the DOI, so this clarifies that a bit.

File: api/resources/views.py [hunk 2]
```diff
@@ -4,18 +4,50 @@
...
+
+    serializer_class = ResourceSerializer
+
+    parser_classes = (JSONAPIMultipleRelationshipsParser, JSONAPIMultipleRelationshipsParserForRegularJSON)
+
+    def get_permissions_proxy(self):
+        try:
+            registration_guid = self.request.data['registration']
+        except KeyError:
+            raise JSONAPIException('Must provide "registration" relationship in payload"')
```
Comment (brianjgeiger) [inline api/resources/views.py:43]:
```suggestion
            raise JSONAPIException('Must provide "registration" relationship in payload')
```

Comment (brianjgeiger) [reply api/resources/views.py:43]: Also, it's a good idea to include a `source` field with the JSONAPIException so we can see what field is problematic without reading the error string. Not that this error is likely to come up runtime after development is complete, but still, it's a good practice.

File: api/resources/permissions.py [hunk 3]
```diff
@@ -41,6 +41,14 @@ def has_permission(self, request, view):
         return proxy_object.has_permission(auth.user, self.REQUIRED_PERMISSIONS[request.method])
 
 
+class ResourceListPermission(ResourcesPermission, permissions.BasePermission):
+    '''Permissions for the top-level ResourceList endpoint.
+
+    ResourceList only supports POST
+    '''
+    REQUIRED_PERMISSIONS = {'POST': 'admin'}
```
Comment (brianjgeiger) [inline api/resources/permissions.py:49]: Can write contributors not add resources?

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
Just a couple of small things and a question

File: api/resources/permissions.py (hunk 3 above)
Comment (jwalz) [reply api/resources/permissions.py:49]: Let's clarify in Scrum today!

**Review by brianjgeiger — APPROVED** [review]

**Pull request merged**